packages/vmakarichev-datagrok_with_cpp/CallWasmWW/wasm/callWasmUtils.js

```diff
@@ -0,0 +1,710 @@
+                break;
+
+            // columns & new columns
+            case INT_COLUMNS_TYPE:             
+            case NEW_INT_COLUMNS_TYPE:
+            case FLOAT_COLUMNS_TYPE:             
+            case NEW_FLOAT_COLUMNS_TYPE: // allocation memory for columns that are created
+                arg.data.buf = module._malloc(arg.data.numOfRows * arg.data.numOfColumns 
+                    * typeMap[type].BYTES_PER_ELEMENT);
+                break;
+
+            // TODO: process other cases and mistakes
+            default:
+                break; 
+        }  
+    }
+    
+    //console.log('inputs after memory allocation:');
+    //console.log(inputs);
+} // allocateMemoryForBuffer
+
+// Get array of values that are put to wasm-function.
+function getArrOfWasmParams(inputs) {
+    let params = [];
+
+    // Process each type of input
+    for(let arg of inputs) {
+        switch (arg.type) {
+
+            // numbers
+            case NUM_TYPE:            
+            case INT_TYPE:
+            case DOUBLE_TYPE:
+                params.push(arg.data);
+                break;
+
+            // column & new column
+            case INT_COLUMN_TYPE:
+            case FLOAT_COLUMN_TYPE:
+            case NEW_FLOAT_COLUMN_TYPE:
+            case NEW_INT_COLUMN_TYPE:
+                params.push(arg.data.buf);                
+                params.push(arg.data.numOfRows);
+                break;
+            
+            // columns & new columns
+            case INT_COLUMNS_TYPE:
+            case NEW_INT_COLUMNS_TYPE:
+            case FLOAT_COLUMNS_TYPE:
+            case NEW_FLOAT_COLUMNS_TYPE: 
+                params.push(arg.data.buf);
+                params.push(arg.data.numOfRows);
+                params.push(arg.data.numOfColumns);
+                break;
+        
+            // TODO: process other cases and mistakes
+            default:
+                break;
+        }
+    }
+
+    return params;
+} // getArrOfWasmParams
+
+// Get array of types that are put to wasm-function.
+function getArrOfWasmTypes(inputs) {
+    let types = [];
+
+    for(let arg of inputs) {
+        switch (arg.type) { // Process each type of input
+
+            // numbers
+            case NUM_TYPE:
+            case INT_TYPE:
+            case DOUBLE_TYPE:
+                types.push(NUMBER);
+                break;
+
+            // column & new column
+            case INT_COLUMN_TYPE:
+            case FLOAT_COLUMN_TYPE:
+            case NEW_FLOAT_COLUMN_TYPE:
+            case NEW_INT_COLUMN_TYPE:
+                types.push(NUMBER);                
+                types.push(NUMBER);
+                break;
+            
+            // columns & new columns
+            case INT_COLUMNS_TYPE:
+            case NEW_INT_COLUMNS_TYPE:
+            case FLOAT_COLUMNS_TYPE:
+            case NEW_FLOAT_COLUMNS_TYPE:
+                types.push(NUMBER);
+                types.push(NUMBER);
+                types.push(NUMBER);
```

Comment:
guess you've created `shiftMap` for these purposes, so you can use the length from there and use it here in the for cycle?

Reply:
Thanks! I'll consider this